# ДЗ №4: Задачи на формулу Байеса и условную вероятность

## Цель: Вспомнить и закрепить понятие условной вероятности и формулы Байеса.

## Задача 1: Путешествие в страну Х

Вы отправляетесь в страну X, которая славится своими мистическими приключениями. Вам предстоит пройти через две опасные локации: лес и горы.

В лесу есть две возможные тропы - тропа A и тропа B. Вы можете выбрать одну из них случайным образом. Вероятность выбрать тропу A равна 0.4, а тропу B - 0.6. После выбора тропы вы перемещаетесь в горы. Вероятность успешно пройти горы, если вы выбрали тропу A, составляет 0.3. Если же вы выбрали тропу B, вероятность успешного прохождения гор увеличивается до 0.7.


**Вопрос:** Какова вероятность успешно пройти оба этапа путешествия (попасть в страну X) и как она зависит от выбора тропы в лесу?

### Дано:

\begin{eqnarray}
P(A) &=& 0.4 \\
P(B) &=& 0.6 \\
P(S/A) &=& 0.3 \\
P(S/B) &=& 0.7
\end{eqnarray}

### Решение:

$$
P(S) = P(A) \cdot P(S/A) + P(B) \cdot P(S/B) = 0.4 \cdot 0.3 + 0.6 \cdot 0.7 = 0.12 + 0.42 = 0.54
$$

**Ответ: до выбора тропы вероятность успеха 54%, после выбора тропы в 40% случаев (априорных подходов) вероятность успеха падает до 30% и в 60% случаев (априорных подходов) возрастает до 70%**

In [1]:
# проверка на вшивость:
round( 0.4 * 0.3 + 0.6 * 0.70, 10) == 0.54

True

## Задача 2: Лабиринт сокровищ

Вы оказались в лабиринте, который славится своими сокровищами. Чтобы добраться до сокровищ, вам нужно пройти через три замка: замок A, замок B и замок C. Каждый замок защищен различными ловушками и подходы к каждому из них различны.

`Замок A`: Первый замок защищен огненными ловушками. Вероятность успешно пройти замок A составляет 0.6.

`Замок B`: После успешного прохождения замка A, вы переходите к замку B. Замок B защищен сложными головоломками. 

* Если вы успешно прошли замок A, вероятность успешно пройти замок B составляет 0.4. 
* Однако, если вы не смогли пройти замок A, то вероятность успешно пройти замок B уменьшается до 0.2.

`Замок C`: После прохождения замка B, вам предстоит столкнуться с последней преградой - замком C. Замок C охраняется могучим воином. 

* Если вы успешно прошли замок B, вероятность успешно пройти замок C составляет 0.7. Однако, если вы не смогли пройти замок B, то вероятность успешно пройти замок C уменьшается до 0.3.

**Вопрос:** Какова вероятность добраться до сокровища (пройти все три замка) и как она зависит от вашего успешного прохождения предыдущих замков?

### Дано:

\begin{eqnarray}
P(A)        &=& 0.6 \\
P(B/A)      &=& 0.4 \\
P(B/\bar A) &=& 0.2 \\
P(C/B)      &=& 0.7 \\
P(C/\bar B) &=& 0.3
\end{eqnarray}

### Решение:
**ориентируясь на вопрос, предполагаем, что для того, чтобы добраться до сокровища, успешное прохождение замков А и В не является обязательным ни в совокупности, ни по отдельности - советую откорректировать условие на "если вы были обнаружены" и "если вы не были обнаружены при прохождении этих замков", а то не понятно как не пройдя замок A или B, можно все же добраться до сокровища!**

**также ориентируясь на вопрос, можно предположить, "что добраться до сокровища" тождественно "успешно пройти замок C", тогда:**

$$
P(C) = P(B) \cdot P(C/B) + P(\bar B) \cdot P(C/\bar B)
$$

но в свою очередь:

\begin{eqnarray}
P(\bar B) &=& 1 - P(B) \\
P(B)      &=& P(A) \cdot P(B/A) + P(\bar A) \cdot P(B/\bar A) \\
P(\bar A) &=& 1 - P(A)
\end{eqnarray}

теперь все переменные известны - подставляем значения вероятностей в обратном порядке:

\begin{eqnarray}
P(\bar A) &=& 1 - 0.6 & & &=& 0.4 \\
P(B)      &=& 0.6 \cdot 0.4 + 0.4 \cdot 0.2 &=& 0.24 + 0.08 &=& 0.32 \\
P(\bar B) &=& 1 - 0.32 & & &=& 0.68 \\
P(C)      &=& 0.32 \cdot 0.7 + 0.68 \cdot 0.3 &=& 0.224 + 0.204 &=& 0.428
\end{eqnarray}

**Ответ часть 1: до начала путешествия по замкам вероятность успеха 42.8%, после прохождения замков A и B в 68% исходных случаев (априорных подходов) вероятность успеха падает до 30% и в 32% исходных случаев (априорных подходов) возрастает до 70%**

In [2]:
# проверка на вшивость:
round(0.68 * 0.3 + 0.32 * 0.70, 10) == 0.428

True

**но вот вопроc: а как корректируются шансы на успех после прохождения только первого замка A - придется также посчитать $P(C/A)$ и $P(C/\bar A)$**

\begin{eqnarray}
P(C/A) &=& \frac{P(A,C)}{P(A)} &=& \frac{(P(A,B,C) + P(A,\bar B,C)}{P(A)} \\
P(C/\bar A) &=& \frac{P(\bar A,C)}{P(\bar A)} &=& \frac{(P(\bar A,B,C) + P(\bar A,\bar B,C)}{P(\bar A)}
\end{eqnarray}

упс, совместное распределение трех случайных величин - этому нас не учили, но уже наработанная интуиция и образование теорфизика в прошлом веке подсказывают мне, что совместное распределение можно набирать последовательными шагами в любом порядке - мы пойдем в порядке прохождения замков, так как именно на этом пути нам заданы условные вероятности задачи:

\begin{eqnarray}
P(A,B,C)           &=&  P(A)\cdot P(B/A) \cdot P(C/B) \\
P(A,\bar B,C)      &=&  P(A)\cdot P(\bar B/A) \cdot P(C/\bar B) \\
P(\bar A,B,C)      &=&  P(\bar A)\cdot P(B/\bar A) \cdot P(C/B) \\
P(\bar A,\bar B,C) &=&  P(\bar A)\cdot P(\bar B/\bar A) \cdot P(C/\bar B)
\end{eqnarray}

можно сократить в аналитическом виде:

\begin{eqnarray}
P(C/A)      &=& P(B/A) \cdot P(C/B)      + P(\bar B/A) \cdot P(C/\bar B) \\
P(C/\bar A) &=& P(B/\bar A) \cdot P(C/B) + P(\bar B/\bar A) \cdot P(C/\bar B)
\end{eqnarray}

неизвестными остались только $P(\bar B/A)$ и $P(\bar B/\bar A)$, но они находятся из тождеств:

\begin{eqnarray}
P(\bar B/A) + P(B/A)           &=& 1 \\
P(\bar B/\bar A) + P(B/\bar A) &=& 1
\end{eqnarray}

продолжаем преобразовывать в аналитической форме:

\begin{eqnarray}
P(C/A)      &=& P(B/A) \cdot P(C/B)      - P(B/A) \cdot P(C/\bar B) + P(C/\bar B) \\
P(C/\bar A) &=& P(B/\bar A) \cdot P(C/B) - P(B/\bar A) \cdot P(C/\bar B) + P(C/\bar B)
\end{eqnarray}

теперь все известно - подставляем:

\begin{eqnarray}
P(C/A)      &=& 0.4 \cdot 0.7 - 0.4 \cdot 0.3 + 0.3 = 0.28 - 0.12 + 0.3 = 0.46 \\
P(C/\bar A) &=& 0.2 \cdot 0.7 - 0.2 \cdot 0.3 + 0.3 = 0.14 - 0.06 + 0.3 = 0.38
\end{eqnarray}

**Ответ часть 2: до начала путешествия по замкам вероятность успеха 42.8%, после прохождения только первого замка A в 60% исходных случаев (априорных подходов) вероятность успеха возрастает до 46% и в 40% исходных случаев (априорных подходов) падает до 38%**

In [3]:
# проверка на вшивость:
round(0.6 * 0.46 + 0.4 * 0.38, 10) == 0.428

True

### ВЫВОД: парадоксальная история:
1) после первого шага (замка) в большинстве исходных случаев прогноз успеха незначительно улучшается,
2) а вот после второго шага в большинстве исходных случаев прогноз будет гораздо хуже исходного!

**да уж, интуиция тут не наработана - надо тупо считать!**

## Задача 3: Заболевание и результаты тестирования

Представим, что у определенной популяции есть редкое заболевание, которое встречается у 1% людей. Существует медицинский тест для определения этого заболевания. Тест имеет высокую точность: вероятность правильно определить наличие заболевания у больного составляет 0.95, а вероятность правильно определить отсутствие заболевания у здорового составляет 0.90.

**Вопрос:** Если у человека положительный тест на заболевание, какова вероятность того, что он действительно болен этим заболеванием?

### Дано:
**обозначения случайных событий:**
* `P  ` = Реальность (Positive)
* `T  ` = Тест (True)
* `T/P` = True Positive (как в ML!)

\begin{eqnarray}
P(P)             &=& 0.01 \\
P(T/P)           &=& 0.95 \\
P(\bar T/\bar P) &=& 0.90
\end{eqnarray}

### Решение:

$$
P(P/T) = \frac{P(T,P)}{P(T)} = \frac{P(P)\cdot P(T/P)}{1-P(\bar T)}
$$

неизвестно $P(\bar T)$:

$$
P(\bar T) = P(\bar T,P) + P(\bar T,\bar P) = P(P)\cdot P(\bar T/P) + P(\bar P)\cdot P(\bar T/\bar P) = 
P(P)\cdot (1-P(T/P)) + (1-P(P))\cdot P(\bar T/\bar P)
$$

теперь все известно - подставляем в обратном порядке:

$$
P(\bar T) = 0.01\cdot(1-0.95) +(1-0.01)\cdot 0.90 = 0.01\cdot 0.05 + 0.99\cdot 0.90 = 0.8915
$$

In [4]:
# проверка на вшивость:
round(0.01*0.05 + 0.99*0.9, 10) == 0.8915

True

$$
P(P/T) = \frac{0.01\cdot 0.95}{1-0.8915} = 0.0875576037
$$

In [5]:
# проверка на вшивость:
round(0.01 * 0.95 / (1 - 0.8915), 10) == 0.0875576037

True

### оценим точность теста 

In [6]:
print('ACCURACY =', (0.01*0.95 + 0.99*0.90))

ACCURACY = 0.9005


### ВЫВОД и Ответ: 
1) хотя accuracy теста 90.05%, вероятность что пациент с положительным тестом реально болен всего 8.76%
2) причина в 10% False Positive результатов на 99% пациентов - они размывают в десять раз 95%-точный прогноз на больных!

**еще раз, интуиция тут не наработана - надо тупо считать!**

## Задача 4: Студенты и предметы

В университете есть два курса по математике: Алгебра и Анализ. Студенты изучают оба предмета, и 70% студентов посещают курс Алгебры, а 30% студентов посещают курс Анализа. Вероятность того, что студент с курса Алгебры сдаст экзамен, составляет 80%, а вероятность того, что студент с курса Анализа сдаст экзамен, составляет 70%.

**Вопрос:** Если студент сдал экзамен по математике. Какова вероятность того, что он изучал курс Алгебры?

### Дано:

**обозначения случайных событий:**
* `A` - ходил на Алгебру
* `B` - ходил на Анализ
* `C` - сдал экзамен по математике

\begin{eqnarray}
P(A)   &=& 0.7 \\
P(B)   &=& 0.3 \\
P(C/A) &=& 0.8 \\
P(C/B) &=& 0.7
\end{eqnarray}

### Решение:

$$
P(A/C) = \frac{P(A,C)}{P(C)} = \frac{P(A)\cdot P(C/A)}{P(C)}
$$

упс: в знаменателе стоит вероятность, которую нетривиально найти алгебраическим перебором!  Более того, ощущение такое, что задача в целом недоопределена, и степеней свободы у системы больше, чем условий в дано.  Однако и в таких задачах есть шанс, что у спрашиваемой величины свободные параметры чудным образом взаимно уничтожатся и/или сократятся.  Так что, поехали опеделять всю систему и ее свободные параметры - если они не сократятся/уничтожатся для искомой величины, то возьмем пару разных и тем покажем, что найдены два разных ответа, удовлетворяющих одинаковым данным - тоже результат.

### Распишем вероятностную систему - для нахождения $P(C)$ и подстановки в Решение выше:
1) множество равноправных элементарных событий $\Omega$ - все студенты универа, сдающие экзамен по математике
2) на нем определены три случайных величины $A$, $B$ и $C$
3) про А и В не известно, что они независимы, поэтому в общем случае $P(A,B)\ne P(A)\cdot P(B)$
4) по факту, С определяется как разноконстантные значения на множествах $(A,B)$, $(A,\bar B)$, $(\bar A,B)$ и $(\bar A, \bar B)$, которые в свою очередь разбивают $\Omega$ на четыре непересекающихся подмножества, а значит вероятность любого события можно получить константной вероятностью его на каждом из четырех - см. пункт ниже 
5) тогда $P(C)$ можно расписать как:

$$
P(C) = P(C/(A,B))\cdot P(A,B) + P(C/(A,\bar B))\cdot P(A,\bar B) + P(C/(\bar A,B))\cdot P(\bar A,B) +P(C/(\bar A,\bar B))\cdot P(\bar A,\bar B)
$$

**когда все болдом, все сливается и трудно визуально видеть, что известно, а что нет, - поэтому переведем все *известные* параметры в *строчные***

поехали параметризировать:

\begin{eqnarray}
P(A,B)      &\equiv& x &\text{(первый параметр)} \\
P(A)        &\equiv& p_a &\text{(просто переобозначим)} \\
P(B)        &\equiv& p_b &\text{(просто переобозначим)} \\
P(A,\bar B)      &=& p_a - x & \\
P(\bar A,B)      &=& p_b - x & \\
P(\bar A,\bar B) &=& 1 - \Bigl( x + (p_a-x) + (p_b-x)\Bigr) = 1 - p_a - p_b + x & \\
& & &\\
P(C/(A,B))  &\equiv& y &\text{(второй параметр)}
\end{eqnarray}

после ввода параметров выражение для $P(C)$ можно переписать как:

$$
P(C) = y\cdot x + P(C/(A,\bar B))\cdot \bigl(p_a-x\bigr) + P(C/(\bar A,B))\cdot \bigl(p_b-x\bigr) + P(C/(\bar A,\bar B))\cdot \bigl(1 - p_b - p_a + x\bigr)
$$

вот теперь визуально видно, что у нас остались три, еще не выраженные через $x$ и $y$, величины: $P(C/(A,\bar B))$, $P(C/(\bar A,B))$ и $P(C/(\bar A,\bar B))$

у нас есть два пока еще не использованных условия данных:

\begin{eqnarray}
P(C/A)  &\equiv& p_{c/a}        & &                                      & \text{(просто переобозначим)} \\
P(C/B)  &\equiv& p_{c/b}        & &                                      & \text{(просто переобозначим)} \\
p_{c/a} &=& \frac{P(C,A)}{P(A)} &=& \frac{P(C,A,B) + P(C,A,\bar B)}{p_a} & = \frac{y\cdot x + P(C/(A,\bar B))\cdot (p_a - x)}{p_a}\\
p_{c/b} &=& \frac{P(C,B)}{P(B)} &=& \frac{P(C,A,B) + P(C,\bar A,B)}{p_b} & = \frac{y\cdot x + P(C/(\bar A,B))\cdot (p_b - x)}{p_b}\\
\end{eqnarray}

два последних легко разрешить относительно $P(C/(A,\bar B))$ и $P(C/(\bar A,B))$:

\begin{eqnarray}
P(C/(A,\bar B))  &=& \frac{p_{c/a}\cdot p_a - y\cdot x}{p_a - x} \\
P(C/(\bar A,B))  &=& \frac{p_{c/b}\cdot p_b - y\cdot x}{p_b - x} \\
\end{eqnarray}

подставляем в выражение для $P(C)$:

$$
P(C) = p_{c/a}\cdot p_a + p_{c/b}\cdot p_b - y\cdot x + P(C/(\bar A,\bar B))\cdot \bigl(1 - p_b - p_a + x\bigr)
$$

и вообще-то, $P(C/(\bar A, \bar B))$ никак не связан условиями задачи, кроме ограничения на значения ВСЕХ вероятностей быть на отрезке $[0,1]$ при сохранении условий задачи

### Замечание 1 по физсмыслу свободных параметров - про халявщиков:
физический смысл этой условной вероятности - сдать экзамен на халяву, то есть, без посещения лекций, поэтому обозначим ее $p_H$

$$
P(C/(\bar A, \bar B)) \equiv p_H \,\,\,\,\,\,\,\,\,\,\,\,\,\,\,\,\,\text{(просто переобозначим)}
$$

заметим, что при строгой системе оценок в университете, эта вероятность должна стремиться к нулю!

### Замечание 2 по физсмыслу свободных параметров - про зубрил:
физический смысл есть и у других свободных параметров: например, $y\equiv P(C/(A,B))$ соответствие лекций программе - это вероятность сдать экзамен, если ходил и на Алгебру и на Анализ, без посещения лекций, поэтому переобозначим его как $p_L$

$$
P(C/(A,B)) \equiv y \equiv p_L \,\,\,\,\,\,\,\,\,\,\,\,\,\,\,\,\,\text{(еще раз переобозначим)}
$$

заметим, что при соответствии программ лекций и экзаменов, эта вероятность должна стремиться к единице!

### Замечание 3 по физсмыслу свободных параметров - "коль сел а-а, то сделаешь и пи-пи":
ну а $x\equiv P(A,B)$ отвечает за независимость выбора предметов студентами, при полностью независимом (нереальное предположение!) он был бы равен $P(A)\cdot P(B) \equiv p_a\cdot p_b$, поэтому лучше ввести корреляционный параметр $\epsilon_{Corr}$ , который будет равен нулю при независимом выборе предметов:

$$
P(A,B) \equiv x \equiv p_a\cdot p_b\cdot(1+\epsilon_{Corr}) \,\,\,\,\,\,\,\,\,\,\,\,\,\,\,\,\,\text{(просто переобозначим)}
$$

заметим, что $\epsilon_{Corr} = 0$ для независимого выбора предмета, что маловероятно!  Однако именно ноль будет естественным дефолтным значением этого свободного параметра

### финальное выражение для вероятности сдать экзамен произвольно выбранному студенту

$$
P(C) = p_{c/a}\cdot p_a + p_{c/b}\cdot p_b - p_L\cdot p_a\cdot p_b\cdot(1+\epsilon_{Corr}) + p_H\cdot \bigl(1 - p_b - p_a + p_a\cdot p_b\cdot(1+\epsilon_{Corr})\bigr)
$$

### далее, чтобы не напахать с цифирью, доверимся питону:

**заведем условия задачи**

In [7]:
p_a = 0.7
p_b = 0.3
p_cVSa = 0.8
p_cVSb = 0.7

**определим функции для вероятностей согласно полученным аналитическим формулам и дефолтным значениям свободных параметров**

In [8]:
def P_C(p_a, p_b, p_cVSa, p_cVSb, p_L=1.0, p_H=0.0, eps_corr=0.0):
    answer  = p_cVSa * p_a + p_cVSb * p_b
    answer -= p_L * p_a * p_b * (1.0 + eps_corr)
    answer += p_H * (1.0 - p_a - p_b + p_a * p_b * (1.0 + eps_corr) )
    return round(answer,10)

def P_AvsC(p_a, p_b, p_cVSa, p_cVSb, p_L=1.0, p_H=0.0, eps_corr=0.0):
    answer = p_a * p_cVSa / P_C(p_a, p_b, p_cVSa, p_cVSb, p_L=p_L, p_H=p_H, eps_corr=eps_corr)
    return round(answer,10)

**Ответ часть 1: - в дефолтных предположениях**

In [9]:
print('\n В предположениях об идеальном образовательном процессе\n',
      '(если везде ходил - то точно сдашь, если никуда не ходил - то точно не сдашь),\n',
      'а также о независимом выборе предметов,\n',
      'вероятность произвольно взятому студенту сдать экзамен равна', 
      P_C(p_a, p_b, p_cVSa, p_cVSb))

print('\n при этом искомая условная вероятность что сдавший экзамен студент ходил на Алгебру равна', 
      P_AvsC(p_a, p_b, p_cVSa, p_cVSb) )


 В предположениях об идеальном образовательном процессе
 (если везде ходил - то точно сдашь, если никуда не ходил - то точно не сдашь),
 а также о независимом выборе предметов,
 вероятность произвольно взятому студенту сдать экзамен равна 0.56

 при этом искомая условная вероятность что сдавший экзамен студент ходил на Алгебру равна 1.0


**Ответ часть 2: - дадим шанс тупым, халявщикам и стремящимся к полным знаниям**

In [10]:
# шанс тупым
p_L_chance = 0.99

# шанс халявщикам
p_H_chance = 0.01

# шанс любителям полных знаний
eps_corr_chance = 0.01

print('\n Однако, если образовательный процесс не полностью идеален\n',
      '(даже если везде ходил - можешь и не сдать, даже если никуда не ходил - шанс есть),\n',
      'а также есть положительная корреляция в выборе предметов по математике,\n',
      'вероятность произвольно взятому студенту сдать экзамен может быть равна', 
      P_C(p_a, p_b, p_cVSa, p_cVSb, p_L=p_L_chance, p_H=p_H_chance, eps_corr=eps_corr_chance))

print('\n при этом искомая условная вероятность что сдавший экзамен студент ходил на Алгебру станет', 
      P_AvsC(p_a, p_b, p_cVSa, p_cVSb, p_H=p_H_chance, eps_corr=eps_corr_chance) )


 Однако, если образовательный процесс не полностью идеален
 (даже если везде ходил - можешь и не сдать, даже если никуда не ходил - шанс есть),
 а также есть положительная корреляция в выборе предметов по математике,
 вероятность произвольно взятому студенту сдать экзамен может быть равна 0.562142

 при этом искомая условная вероятность что сдавший экзамен студент ходил на Алгебру станет 0.9999625014


**Ответ часть 3: - дадим шанс тупым, халявщикам и ленивым**

In [11]:
# шанс тупым
p_L_chance = 0.99

# шанс халявщикам
p_H_chance = 0.01

# шанс ленивым
eps_corr_chance = - 0.01

print('\n Однако, если образовательный процесс не полностью идеален\n',
      '(даже если везде ходил - можешь и не сдать, даже если никуда не ходил - шанс есть),\n',
      'а также есть отрицательная корреляция в выборе предметов по математике,\n',
      'вероятность произвольно взятому студенту сдать экзамен может быть равна', 
      P_C(p_a, p_b, p_cVSa, p_cVSb, p_L=p_L_chance, p_H=p_H_chance, eps_corr=eps_corr_chance))

print('\n при этом искомая условная вероятность что сдавший экзамен студент ходил на Алгебру станет', 
      P_AvsC(p_a, p_b, p_cVSa, p_cVSb, p_H=p_H_chance, eps_corr=eps_corr_chance) )


 Однако, если образовательный процесс не полностью идеален
 (даже если везде ходил - можешь и не сдать, даже если никуда не ходил - шанс есть),
 а также есть отрицательная корреляция в выборе предметов по математике,
 вероятность произвольно взятому студенту сдать экзамен может быть равна 0.566258

 при этом искомая условная вероятность что сдавший экзамен студент ходил на Алгебру станет 0.9925927764


### ВЫВОДЫ:
1) задача не доопределена - у системы три степени свободы, ограниченные на краях условием пренадлежности ВСЕХ мыслимых вероятностей отрезку $[0,1]$.  Выше мы показали, что есть разные наборы свободных параметров, для которых ответ задачи существует и различен

2) задача становится полностью определенной, если добавить естественные допущения:
 * что без посещения хотя бы одного предмета сдать математику нельзя (например, система баллов в оценке не позволяет)
 * что при посещении всех предметов сдача гарантирована (например, автомат от лектора)
 * что выбор предметов независим (семантически это маловероятно, разве что, вырожденный случай, например, лентяи точно сбалансированы любителями знаний, но тем не менее, это самое простое предположение)

3) в последнем случае ответ задачи равен ровно 1.0 

In [12]:
from datetime import datetime
print('LAST RUN on', str(datetime.now()).split('.')[0])

LAST RUN on 2025-12-10 17:22:51
